Lighting rewrite

mloc opened the pull request:
Coloured, animated, object-based lighting system with fancy luminescence. (lambertian shading)

Other changes:
- Lighting subarea system is gone
- Lights in areas that don't require power will now automatically turn on
- More lightweight camera cache invalidation
- "simulated" var for atoms
- lights in shuttles!

Changes from tg DAL:
- Lighting is done using objects instead of subareas.
- Animated transitions.
- Full colours with mixing.
- Support for lights on shuttles.
- Code:
  - Instead of one flat luminosity var, light is represented by 3 atom vars:
    - light_range; range in tiles of the light, used for calculating falloff,
    - light_power; multiplier for the brightness of lights,
    - light_color; hex string representing the RGB colour of the light.
  - SetLuminosity() is now set_light() and takes the three variables above.
    - Variables can be left as null to not update them.
  - SetOpacity() is now set_opacity().
  - Areas have luminosity set to 1 permanently, no hard-lighting.
  - Objects inside other objects can have lights and they properly affect the turf. (flashlights)
  - area/master and area/list/related have been eviscerated since subareas aren't needed.

todo:
- [x] Fix bugs:
  - [x] Ghost light sources
  - [x] Space/shuttle areas
  - [x] Replace instance luminosity with light_power/range
- [x] Optimize:
  - [x] apply_lum() tweaking
  - [ ] maybe pooling for light sources
- [x] Destroy area.related/master
- [x] Prettying up:
  - [x] Tweak light power for station lights
  - [x] Starlight
  - [x] Mood lighting for machines
    - Computers
    - Space heaters?
    - Atmos machinery?
- [x] Finalization:
  - [x] Documentation
  - [x] Fix file organization
  - [x] Pictures/vids for PR
  - [x] Changelog
  - [x] Rebase/clean history

Vids stolen from ParadiseSS13/Paradise#998
http://puu.sh/hIn4e/78df641df2.mp4
http://puu.sh/hIn9r/c3200e30fd.mp4

![](https://i.imgur.com/Ua2sR1g.png)
![](https://i.imgur.com/7BOKgTE.png)
![](https://i.imgur.com/v93G2lF.png)

comma commented on code/game/machinery/flasher.dm:
```diff
@@ -27,16 +27,16 @@
 /obj/machinery/flasher/New()
 	sleep(4)					//<--- What the fuck are you doing? D=
-	src.sd_SetLuminosity(2)
+	src.sd_set_light(2)
```
Merciless find&replace in this file

comma commented on code/game/machinery/igniter.dm:
```diff
@@ -66,10 +66,10 @@
 
 		icon_state = "[base_state]"
-//		src.sd_SetLuminosity(2)
+//		src.sd_set_light(2)
```
More of grep butchery

ZomgPonies commented:
Pretty please keep working on this :+1:

mloc committed:
remove the sub-area system

Removes area/list/related and area/master.
They were only used by the old lighting subarea system.

Signed-off-by: Mloc <colmohici@gmail.com>

mloc committed:
lights in areas that don't require power will now automatically turn on

Signed-off-by: Mloc <colmohici@gmail.com>

mloc committed:
more lightweight camera caching

no loops

Signed-off-by: Mloc <colmohici@gmail.com>

mloc committed:
lighting map changes

Signed-off-by: Mloc <colmohici@gmail.com>

mloc committed:
changelog for lighting rewrite

Signed-off-by: Mloc <colmohici@gmail.com>

mloc committed:
simulated var for atoms, used to filter explosions and other actions

kind of a stronger anchored var, used by lighting overlays

Signed-off-by: Mloc <colmohici@gmail.com>

mloc committed:
mood lighting for computers

Signed-off-by: Mloc <colmohici@gmail.com>

mloc committed:
dynamic fire colours

Fire colour + light colour changes with temperature.

Signed-off-by: Mloc <colmohici@gmail.com>

mloc committed:
new object-based lighting system

Signed-off-by: Mloc <colmohici@gmail.com>

Mloc commented:
ok this is done

comma commented on code/modules/clothing/clothing.dm:
```diff
@@ -308,18 +290,6 @@ BLIND     // can't see anything
 		var/mob/living/carbon/human/H = user
 		H.update_inv_head()
 
-/obj/item/clothing/head/equipped(mob/user)
```
These aren't needed anymore because light's in object instead of carrier, right?

comma commented:
:+1:

Fox-McCloud commented:
Question, Mloc.

This makes darkness absolute and the species vision in the dark completely and totally worthless; if you're in a dark area, you can't see jack squat, even if you're, Unathi or Tajaran; previously humans could see 2 tiles in the dark and anything past that was absolute darkness. Unathi could see (IIRC) 3--and Tajarans had perfect vision range in the dark. Things were still very dark, for them, but they could still definitely see things.

Fox-McCloud commented on code/game/objects/explosion.dm:
```diff
@@ -87,8 +87,8 @@ proc/explosion(turf/epicenter, devastation_range, heavy_impact_range, light_impa
-		for(var/turf/T in range(epicenter, max_range))
-			var/dist = cheap_pythag(T.x - x0,T.y - y0)
+		for(var/turf/T in trange(max_range, epicenter))
+			var/dist = sqrt((T.x - x0)**2 + (T.y - y0)**2)
```
@Mloc 

I realize this is hugely late, but I just now noticed this. What exactly was the reason for changing this and what are the advantages of it over cheap_pythag?

GinjaNinja32 replied:
If I had to guess which was faster, I'd guess `sqrt` and `**`, since they're implemented in C++ instead of DM. Just a guess, though, I might do a bit of benchmarking later, unless someone appears with some.

Fox-McCloud replied:
There is a very distinctive difference in the shape of the explosions; it's more square than round with this formula, but---if it is faster, well, it's probably "better" then.

GinjaNinja32 replied:
It should still be round, they're both computing a distance from a central turf.

GinjaNinja32 replied:
https://gist.github.com/GinjaNinja32/5747d5b497f74c725982

On my machine:
>test_cheappythag(500) in 5 ds
>test_sqrt(500) in 3 ds
>test_cheappythag(1000) in 20 ds
>test_sqrt(1000) in 10 ds
>test_cheappythag(1500) in 45 ds
>test_sqrt(1500) in 21 ds
>test_cheappythag(5000) in 516 ds
>test_sqrt(5000) in 230 ds

mwerezak replied:
So much for "cheap"

Fox-McCloud replied:
Holy damn, it's half the cost.

PsiOmegaDelta replied:
GinjaNinja32 did some some performance testing, cannot find the PR with the comment and data right now though, and this about twice as fast as cheap_pythag().
The 'cheap' procs were completely removed in #10066 (though I should perhaps have suggested making it a macro at the time).
